p2p/protocol/identify/id.go

```diff
@@ -308,13 +346,30 @@ func (ids *IDService) populateMessage(mes *pb.Identify, c network.Conn) {
 	// "public" address, at least in relation to us.
 	mes.ObservedAddr = c.RemoteMultiaddr().Bytes()
 
-	// set listen addrs, get our latest addrs from Host.
-	laddrs := ids.Host.Addrs()
-	mes.ListenAddrs = make([][]byte, len(laddrs))
-	for i, addr := range laddrs {
-		mes.ListenAddrs[i] = addr.Bytes()
+	if usePeerRecords {
+		// Generate a signed peer record containing our listen addresses and
+		// send it instead of unsigned addrs
+		rec := ids.peerRecordManager.LatestRecord()
+		if rec == nil {
+			log.Errorf("latest PeerRecord does not exist. identify message incomplete!")
+		} else {
+			recBytes, err := rec.Marshal()
```

Comment:
Hm. Can we pre-marshal this?

Reply:
Sure. I'll do this in the next PR once we have a `GetRecord` method.

Reply:
https://github.com/libp2p/go-libp2p/issues/897